### PR TITLE
[MM-52207] Grammarly plugin not detected in Mattermost desktop

### PR DESCRIPTION
**Pull request by hasancankucuk**
#### Summary
The aria-label attribute has been modified, and it is impacting the text-area property. Grammarly couldn't detect the issue since the text-area was affected.

#### Ticket Link
Fixes: https://github.com/mattermost/mattermost/issues/22970
Jira: https://mattermost.atlassian.net/browse/MM-52207

#### Screenshots
|before | after |

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx:468]: I am not sure if this is exactly how this issue is supposed to be solved, honestly, I don't have the solution either. Please note that the issue is currently doesn't appear to be in Mattermost run in either Chrome, Firefox, but in Desktop app only.

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Comment by M-ZubairAhmed** [comment]
@hasancankucuk i have updated the original issue can you please check

File: webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx
Comment (hasancankucuk) [reply webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx:468]: @M-ZubairAhmed After testing the changes I made to Mattermost on MacOS, it appears that the problem has been successfully resolved. I have attached screenshots for reference. The reason behind this solution is that when aria-label is added, Grammarly treats the textarea as a string, making it unable to detect. To address this issue, I included a label to prevent the textarea from being treated as a string, ensuring both accessibility is maintained and Grammarly detects it correctly. I conducted comprehensive testing using VoiceOver and Grammarly version 1.30.2.0 to validate the improvements


**
| before_desktop | after_desktop  |**

**Comment by hasancankucuk** [comment]
@M-ZubairAhmed While looking for another solution to the problem, I found useful information in React's official documentation. The accessibility information provided by React matches the solution I made. Here is the link for your reference if you want to have a look.

telegram-cloud-photo-size-4-5949373043719650390-y


Documentation Link: https://react.dev/reference/react-dom/components/textarea

File: webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx:466]: These IDs need to be unique because there can actually be multiple `AdvancedTextEditor`s on screen at once: one in the centre channel for posts and one in the RHS for comments. `props.location` should be different between those two, so could you change this to 
```js
`advancedTextEditorLabel-${location}`
```
so that both of those have different IDs?

I noticed React now has a `useId` hook to use for this case, but unfortunately, we can't use it here because we're not on React 18 yet.

File: webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx
Comment (hmhealey) [inline webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx:467]: Instead of passing a style object, could you change this to use the `sr-only` CSS class? That's typically what we do for elements like this

File: webapp/channels/src/components/autosize_textarea.tsx
Comment (hmhealey) [inline webapp/channels/src/components/autosize_textarea.tsx:161]: Same comments here regarding the ID needing to be unique and that this should use the `sr-only` class. This component already has a `this.props.id`, so we can make the label's ID `id={id + '-label'}` to match the other IDs in this file

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR! I haven't been able to test the PR yet since I don't have Grammarly installed myself, but I do have some feedback first before I can get someone else to test this.

In addition to my comments on the code, could you also ensure the unit tests for the web app run by running `npm test -w channels` from the `webapp` folder? Based on the CI results, those are currently failing because the snapshots of the `AutosizeTextarea` component have changed

**Comment by devinbinnie** [comment]
@hasancankucuk Are you able to address the above comments? Also looks like there is a snapshot failing.

**Comment by hasancankucuk** [comment]
@hmhealey @devinbinnie Hi, I was having some technical issues so I couldn't make the changes, I'll fix them ASAP

**Comment by devinbinnie** [comment]
@hasancankucuk Any luck on getting the issues fixed?

**Comment by hasancankucuk** [comment]
@devinbinnie I made the updates suggested by @hmhealey and updated the snapshot.

**Comment by devinbinnie** [comment]
> @devinbinnie I made the updates suggested by @hmhealey and updated the snapshot.

Thanks! We will have a look.

**Comment by raky6666** [comment]
When will the release take place?

**Comment by hmhealey** [comment]
Removing stale since this is blocked on us at the moment, and updating it again to hopefully fix the EE tests

/update-branch

**Review by hmhealey — APPROVED** [review]
I tested this locally on a Windows 11 VM, and this fix worked for me. Thanks for the PR!

**Review by devinbinnie — APPROVED** [review]
All LGTM! Thanks @hasancankucuk!

**Review by yasserfaraazkhan — CHANGES_REQUESTED** [review]
@hasancankucuk can you check the A11y tests failing, you can find the report [here](https://automation-dashboard.vercel.app/cycle/374335) 

in case pr changes are causing it

**Comment by hmhealey** [comment]
Looking through those E2E tests, these ones look like they're actually related to this PR since they're looking for the aria-label of the post textbox:

- `e2e-test/cypress/tests/integration/channels/accessibility/accessibility_nav_diff_regions_spec.js`
- `e2e-test/cypress/tests/integration/channels/enterprise/accessibility/accessibility_input_fields_spec.js`
- `e2e-test/cypress/tests/integration/channels/files_and_attachments/upload_files_spec.js`

The rest of them appear to be unrelated.

There's [docs for the E2E tests here](https://developers.mattermost.com/contribute/more-info/webapp/e2e-testing/), and if you need help getting those working up, please let us know

**Comment by coltoneshaw** [comment]
I don't see Grammarly detecting this on the desktop application for MacOS. It does seem to work in web.


Screenshot 2023-10-03 at 4 28 56 PM

Screenshot 2023-10-03 at 4 30 44 PM

**Comment by devinbinnie** [comment]
@hasancankucuk Can you look at the conflicts and the failing test and see if there are any issues? Otherwise this looks good to merge.

@coltoneshaw Is it just Desktop App that isn't working? If so we can create a separate issue to see if Electron is blocking it somehow.

**Comment by devinbinnie** [comment]
> @hasancankucuk Can you look at the conflicts and the failing test and see if there are any issues? Also, are you testing on the Mac App Store version of the Desktop App?

Was able to have it work on the regular Desktop App build but there's an issue with the Mac App Store build. I'll have to fix that separately, so we can still go ahead and merge this when ready.

**Comment by hasancankucuk** [comment]
@devinbinnie hi, i'll update the failing test asap. Also I didn't try with the macos version.

**Comment by hasancankucuk** [comment]
@devinbinnie when i run cypress tests locally, accessibility_nav_diff_regions_spec.js and upload_files_spec.js are passing the tests. Just so you know, I'm attaching accessibility_nav_diff_regions_spec.js screenshots. And if I update the tests it's failing. I don't understand what I am missing. Also when I try to run accessibility_input_fields_spec.js it fails because of the trial-license API I am also attaching its screenshot.

Ekran Resmi 2023-10-18 20 25 28
Ekran Resmi 2023-10-18 20 21 05
Ekran Resmi 2023-10-18 20 20 08

**Comment by devinbinnie** [comment]
> @devinbinnie when i run cypress tests locally, accessibility_nav_diff_regions_spec.js and upload_files_spec.js are passing the tests. Just so you know, I'm attaching accessibility_nav_diff_regions_spec.js screenshots. And if I update the tests it's failing. I don't understand what I am missing. Also when I try to run accessibility_input_fields_spec.js it fails because of the trial-license API I am also attaching its screenshot.
> 
> Ekran Resmi 2023-10-18 20 25 28 Ekran Resmi 2023-10-18 20 21 05 Ekran Resmi 2023-10-18 20 20 08

cc @yasserfaraazkhan

File: webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx:433]: please use react intl methods here

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx
Comment (hasancankucuk) [reply webapp/channels/src/components/advanced_text_editor/advanced_text_editor.tsx:433]: I don't think it's a problem specific to the desktop application. When I tested it on the master branch in Safari, it still couldn't detect Grammarly Mattermost as in the screenshots. 

Ekran Resmi 2023-10-24 23 18 09
Ekran Resmi 2023-10-24 23 18 13
Ekran Resmi 2023-10-24 23 17 13
Ekran Resmi 2023-10-24 23 17 05

**Comment by hasancankucuk** [comment]
@devinbinnie Grammarly is now detecting Mattermost Desktop v5.5.1 both local build and the dmg downloaded from https://docs.mattermost.com/collaborate/install-desktop-app.html . But on the App Store version, it's still not working. On Grammarly's support page, there's info about this situation. I've also checked Mattermost Desktop's electron version and it currently uses version 26.2.1. I think the issue and pull request can be closed. I'm also attaching a screenshot and a link to Grammarly's support page.

https://support.grammarly.com/hc/en-us/articles/4428282986893-Grammarly-for-Mac-doesn-t-work-in-certain-apps-for-Mac
image

**Comment by devinbinnie** [comment]
> @devinbinnie Grammarly is now detecting Mattermost Desktop v5.5.1 both local build and the dmg downloaded from https://docs.mattermost.com/collaborate/install-desktop-app.html . But on the App Store version, it's still not working. On Grammarly's support page, there's info about this situation. I've also checked Mattermost Desktop's electron version and it currently uses version 26.2.1. I think the issue and pull request can be closed. I'm also attaching a screenshot and a link to Grammarly's support page.
> 
> https://support.grammarly.com/hc/en-us/articles/4428282986893-Grammarly-for-Mac-doesn-t-work-in-certain-apps-for-Mac image

Y

> @devinbinnie Grammarly is now detecting Mattermost Desktop v5.5.1 both local build and the dmg downloaded from https://docs.mattermost.com/collaborate/install-desktop-app.html . But on the App Store version, it's still not working. On Grammarly's support page, there's info about this situation. I've also checked Mattermost Desktop's electron version and it currently uses version 26.2.1. I think the issue and pull request can be closed. I'm also attaching a screenshot and a link to Grammarly's support page.
> 
> https://support.grammarly.com/hc/en-us/articles/4428282986893-Grammarly-for-Mac-doesn-t-work-in-certain-apps-for-Mac image

Yeah, this is unfortunately the case for a lot of things. In order to be in the App Store, macOS requires that an app be placed within its App Sandbox that restricts access to a lot of functionality, unless specifically granted. This probably means that Grammarly won't work with our app in the App Store, so this change is the best we can do.

**Comment by devinbinnie** [comment]
@M-ZubairAhmed Are we okay to approve this?

**Comment by M-ZubairAhmed** [comment]
Foloww

> @M-ZubairAhmed Are we okay to approve this?

Following the above discussion i think this is the maximum we can do from outside. Let me rope in @yasserfaraazkhan to test if this fix works for the following

- Webapp in MacOS/Windows/Ubuntu Chrome
- Webapp in MacOS/Windows/Ubuntu Firefox
- Webapp in MacOS Safari
- Webapp in MacOS/Windows/Ubuntu Edge

Please check with and without the above fix.

**Comment by M-ZubairAhmed** [comment]
@hasancankucuk please fix the CI for the preview url to get generated

**Pull request closed without merging**
